backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/InventoryLineRecordService.java

```diff
@@ -405,6 +423,64 @@ public HUAggregationType computeHUAggregationType(
 		return computeHUAggregationType(inventoryLine, docBaseAndSubType);
 	}
 
+	public void setQtyBookedFromStorage(@NonNull final I_M_InventoryLine inventoryLine)
+	{
+		final Optional<Quantity> bookedQty = getBookedQtyFromStorage(inventoryLine);
+
+		if (bookedQty.isPresent())
+		{
+			if (bookedQty.get().getUomId().getRepoId() != inventoryLine.getC_UOM_ID())
+			{
+				//this should never happen as inventoryLineRecordService.getBookedQtyFromStorage() returns the qty in the inventory line's uom.
+				throw new QuantitiesUOMNotMatchingExpection("Booked and counted quantities don't have the same UOM!")
+						.appendParametersToMessage()
+						.setParameter("InventoryLineUOMID", inventoryLine.getC_UOM_ID())
+						.setParameter("BookedQtyUOMID", bookedQty.get().getUomId());
+			}
+
+			inventoryLine.setQtyBook(bookedQty.get().toBigDecimal());
+		}
+		else
+		{
+			inventoryLine.setQtyBook(BigDecimal.ZERO);
+		}
+	}
+
+
+	@NonNull
+	private Optional<Quantity> getBookedQtyFromStorage(@NonNull final I_M_InventoryLine inventoryLineRecord)
```

Comment:
AFAIS  it doesn't work like this; the method doesn't cover the case of multiple HUs for one line.
Rough sketch of how to do it instead:
* change the method to work with `de.metas.handlingunits.inventory.InventoryLine` instead of `I_M_InventoryLine`
  * maybe when it's not a dirty record we work with, the whole code can be moved out of this dirty service
* use `inventoryLine.getInventoryLineHUs()` to get the `n` HUs a line can have
* you probably need IHandlingUnitDAO to load the respective I_M_HUs; (maybe there is an easy way to load them all at once)
* use `IHandlingUnitsBL.getStorageFactory()` and IHUStorageFactory.getHUProductStorages() to get the quantities.
* maybe you don't need `uomConversionBL` to convert them to the inventory line's UOM..might be that's done by the repository (need to check)
* :exclamation: By the way: check out how the processes from today's book-club do it. maybe you can simply extract and/or reuse that code.
* fun fact: your change of `InventoryRepository.retrieveLineRecords` to `public` triggered all these thoughts..opening a nice API like that to server raw database records was kind of a smell :-D
* :exclamation: 
  * AFAIU InventoryLine is not really the aggregate root; Inventory is..
  * in the `M_InventoryLine` MI you need to update a single line though..not sure that to do
  * ..maybe let InventoryRepository load an inventory with just the line in question?
  * the good news is that AFAIK @teosarca will be around tomorrow for consultation
  * ...did i mention that there *might* be some code already doing all this?

Reply:
🤦 yep, you're right, I didn't put much thought into it as it was more like a quick fix 🙃, thx for raising the issue I'll check that process and will try to find a way to reuse the code.